### PR TITLE
Storage map iter

### PR DESCRIPTION
**Pull request by dvc94ch**
Completes #147

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -185,6 +195,42 @@ impl<T: Runtime> Clone for Client<T> {
...
+    start_key: Option<StorageKey>,
+    keys: Vec<StorageKey>,
+}
+
+impl<T: Runtime, F: Store<T>> KeyIter<T, F> {
+    /// Returns the next key value pair from a map.
+    pub async fn next(&mut self) -> Result<Option<(StorageKey, F::Returns)>, Error> {
+        loop {
+            if let Some(key) = self.keys.pop() {
+                if let Some(value) = self.client.rpc.storage(&key, self.hash).await? {
```
Comment (ascjones) [inline src/lib.rs:218]: What about `self.client.rpc.query_storage` with the whole batch of keys for a single roundtrip?

File: src/lib.rs [hunk 2]
```diff
@@ -185,6 +195,42 @@ impl<T: Runtime> Clone for Client<T> {
...
+    _marker: PhantomData<F>,
+    count: u32,
+    hash: Option<T::Hash>,
+    start_key: Option<StorageKey>,
+    keys: Vec<StorageKey>,
+}
+
+impl<T: Runtime, F: Store<T>> KeyIter<T, F> {
+    /// Returns the next key value pair from a map.
+    pub async fn next(&mut self) -> Result<Option<(StorageKey, F::Returns)>, Error> {
```
Comment (ascjones) [inline src/lib.rs:215]: I'm wondering whether it's worth making this `impl Stream` too? Then you could do e.g. `client.account_iter().collect::<Vec<_>>().await`

File: src/lib.rs [hunk 3]
```diff
@@ -200,45 +246,59 @@ impl<T: Runtime> Client<T> {
         &self.metadata
     }
 
-    /// Fetch a StorageKey with default value.
-    pub async fn fetch_or_default<F: Store<T>>(
+    /// Fetch a StorageKey an optional storage key.
```
Comment (ascjones) [inline src/lib.rs:249]: I realise this comment was already there but it doesn't make any sense

File: src/lib.rs (hunk 2 above)
Comment (dvc94ch) [reply src/lib.rs:215]: Would be nice, but it seems terribly annoying to do. Maybe I'm missing a simple solution, but basically once you go from manually implemented futures and streams to async syntax there isn't an easy way back.

Comment (ascjones) [reply src/lib.rs:216]: Yeah. I can't think of a nice and elegant way to do it - maybe a future PR

File: src/lib.rs [hunk 4]
```diff
@@ -185,6 +196,53 @@ impl<T: Runtime> Clone for Client<T> {
...
+    start_key: Option<StorageKey>,
+    buffer: Vec<(StorageKey, StorageData)>,
+}
+
+impl<T: Runtime, F: Store<T>> KeyIter<T, F> {
+    /// Returns the next key value pair from a map.
+    pub async fn next(&mut self) -> Result<Option<(StorageKey, F::Returns)>, Error> {
+        loop {
+            if let Some((k, v)) = self.buffer.pop() {
+                return Ok(Some((k, Decode::decode(&mut &v.0[..])?)))
```
Comment (4meta5) [inline src/lib.rs:219]: @dvc94ch why don't we decode the key here? It is a raw `sp_core::sp_store::StorageKey`, but I'm pushing onto a vector `Vec<(T::BountyId, BountyState<T>)>`. I guess I can manually decode it...

**Pull request merged**